### ISSUE TITLE
How to change the expression when using the faceid model

### ISSUE THREAD

**Issue by guijuzhejiang** [tencent-ailab/IP-Adapter#214]
Great job, I used the faceid model and the characters are very well reproduced. But I also found a problem, probably because of face embedding, the expression of the reference person is highly restored, even if I write different expression ptompts in prompts, I can't change the expression as well, is there any way to change the expression of the reference person?

**Comment by xiaohu2015** [comment]
faceid model should be able to change expressions, faceid plus model maybe not, but you can try faceid plus v2 to achieve that （use lower weight）

**Comment by YZBPXX** [comment]
> faceid model should be able to change expressions, faceid plus model maybe not, but you can try faceid plus v2 to achieve that （use lower weight）


I visualized the face token maps for the faceid and faceid_plus model. From the attention maps, it appears that faceid_plus controls more accurately. Can I merge the visualization code into the main branch? It would simplify visualizing face token maps for all models you release.
<img width="961" alt="截屏2024-01-03 20 41 27" src="https://github.com/tencent-ailab/IP-Adapter/assets/57928886/fee99ccf-88d6-4ce6-aef4-bf4816185b59">
<img width="967" alt="截屏2024-01-03 20 41 52" src="https://github.com/tencent-ailab/IP-Adapter/assets/57928886/19f55e2c-66b4-4bee-9588-a19ad53ec256">

**Comment by xiaohu2015** [comment]
thanks a lot

**Comment by rafstahelin** [comment]
How does one visualise the face embeddings?Best, Raf On 3 Jan 2024, at 13:42, YZBPXX ***@***.***> wrote:﻿

faceid model should be able to change expressions, faceid plus model maybe not, but you can try faceid plus v2 to achieve that （use lower weight）

I visualized the face token maps for the faceid and faceid_plus model. From the attention maps, it appears that faceid_plus controls more accurately. Can I merge the visualization code into the main branch? It would simplify visualizing face token maps for all models you release.
2024-01-03.20.41.27.png (view on web)
2024-01-03.20.41.52.png (view on web)

**Comment by YZBPXX** [comment]
> How does one visualise the face embeddings?Best, Raf On 3 Jan 2024, at 13:42, YZBPXX ***@***.***> wrote:﻿ faceid model should be able to change expressions, faceid plus model maybe not, but you can try faceid plus v2 to achieve that （use lower weight） I visualized the face token maps for the faceid and faceid_plus model. From the attention maps, it appears that faceid_plus controls more accurately. Can I merge the visualization code into the main branch? It would simplify visualizing face token maps for all models you release. 2024-01-03.20.41.27.png (view on web) 2024-01-03.20.41.52.png (view on web) —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>
> 如何可视化面部嵌入？Best， Raf 2024 年 1 月 3 日 13：42，YZBPXX ***@***.***>写道：
> 
> FaceID 模型应该能够更改表情，FaceID Plus 模型可能不能，但您可以尝试 FaceID Plus v2 来实现（使用更轻的权重）
> 
> 我可视化了 faceid 和 faceid_plus 模型的人脸令牌映射。从注意力图来看，faceid_plus控制似乎更准确。我可以将可视化代码合并到主分支中吗？它将简化您发布的所有模型的人脸令牌地图的可视化。
> 2024-01-03.20.41.27.png （在网络上查看）
> 2024-01-03.20.41.52.png （在网络上查看）
> 
> - 直接回复此电子邮件，在 GitHub 上查看或取消订阅。您收到此消息是因为您订阅了此线程。邮件 ID：***@***.***>

You can browse the recently updated code in the 'visual_attnmap.ipynb' notebook.

**Comment by YZBPXX** [comment]
> How does one visualise the face embeddings?Best, Raf On 3 Jan 2024, at 13:42, YZBPXX ***@***.***> wrote:﻿ faceid model should be able to change expressions, faceid plus model maybe not, but you can try faceid plus v2 to achieve that （use lower weight） I visualized the face token maps for the faceid and faceid_plus model. From the attention maps, it appears that faceid_plus controls more accurately. Can I merge the visualization code into the main branch? It would simplify visualizing face token maps for all models you release. 2024-01-03.20.41.27.png (view on web) 2024-01-03.20.41.52.png (view on web) —Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

You can browse the recently updated code in the 'visual_attnmap.ipynb' notebook.

**Comment by juancopi81** [comment]
This is great! Would the get_net_attn_map function work for the ip_adapter_plus? (no faceid)

**Comment by YZBPXX** [comment]
> This is great! Would the get_net_attn_map function work for the ip_adapter_plus? (no faceid)
I just submitted a pull request. You just need to use the following three lines of code as you would use faceid
```python
pipe.unet = register_cross_attention_hook(pipe.unet)
attn_maps = get_net_attn_map((768, 512))
attn_hot = attnmaps2images(attn_maps)
```

**Comment by YZBPXX** [comment]
> This is great! Would the get_net_attn_map function work for the ip_adapter_plus? (no faceid)


I just submitted a pull request. You just need to use the following three lines of code as you would use faceid
```python
pipe.unet = register_cross_attention_hook(pipe.unet)
attn_maps = get_net_attn_map((768, 512))
attn_hot = attnmaps2images(attn_maps)
```

**Comment by juancopi81** [comment]
That's great @YZBPXX Thanks. A final question, in get_nett_attn_map what batch_size means here? Is it related with the pos. neg. conditions or with the number of images?

**Comment by YZBPXX** [comment]
> That's great @YZBPXX Thanks. A final question, in get_nett_attn_map what batch_size means here? Is it related with the pos. neg. conditions or with the number of images?


Yes, currently, I have only considered the case of num_samples=1, batch_size=2 (generate one image requires sampling twice). If generating multiple samples at once, it may throw an error, and some modifications need to be made. I will make modifications later to support num_samples > 1. If you have any more questions, I'd be happy to answer them for you.

**Comment by juancopi81** [comment]
Thanks a lot @YZBPXX! I was looking at my shapes of the attn_maps and found:

self.attn_map.shape torch.Size([2, 8, 4096, 16])
self.attn_map.shape torch.Size([2, 8, 4096, 16])
self.attn_map.shape torch.Size([2, 8, 1024, 16])
self.attn_map.shape torch.Size([2, 8, 1024, 16])
self.attn_map.shape torch.Size([2, 8, 256, 16])
self.attn_map.shape torch.Size([2, 8, 256, 16])
self.attn_map.shape torch.Size([2, 8, 64, 16])
self.attn_map.shape torch.Size([2, 8, 256, 16])
self.attn_map.shape torch.Size([2, 8, 256, 16])
self.attn_map.shape torch.Size([2, 8, 256, 16])
self.attn_map.shape torch.Size([2, 8, 1024, 16])
self.attn_map.shape torch.Size([2, 8, 1024, 16])
self.attn_map.shape torch.Size([2, 8, 1024, 16])
self.attn_map.shape torch.Size([2, 8, 4096, 16])
self.attn_map.shape torch.Size([2, 8, 4096, 16])
self.attn_map.shape torch.Size([2, 8, 4096, 16])

Just to be sure, the last dimension (16) is the features were the image is projected, so for the attention maps, I could average them?

**Comment by YZBPXX** [comment]
> Thanks a lot @YZBPXX! I was looking at my shapes of the attn_maps and found:
> 
> self.attn_map.shape torch.Size([2, 8, 4096, 16]) self.attn_map.shape torch.Size([2, 8, 4096, 16]) self.attn_map.shape torch.Size([2, 8, 1024, 16]) self.attn_map.shape torch.Size([2, 8, 1024, 16]) self.attn_map.shape torch.Size([2, 8, 256, 16]) self.attn_map.shape torch.Size([2, 8, 256, 16]) self.attn_map.shape torch.Size([2, 8, 64, 16]) self.attn_map.shape torch.Size([2, 8, 256, 16]) self.attn_map.shape torch.Size([2, 8, 256, 16]) self.attn_map.shape torch.Size([2, 8, 256, 16]) self.attn_map.shape torch.Size([2, 8, 1024, 16]) self.attn_map.shape torch.Size([2, 8, 1024, 16]) self.attn_map.shape torch.Size([2, 8, 1024, 16]) self.attn_map.shape torch.Size([2, 8, 4096, 16]) self.attn_map.shape torch.Size([2, 8, 4096, 16]) self.attn_map.shape torch.Size([2, 8, 4096, 16])
> 
> Just to be sure, the last dimension (16) is the features were the image is projected, so for the attention maps, I could average them?




the last dimension(16) is the number of 16 face tokens. If you average them, it should yield the combined result of the 16 tokens. I have attempted to do this, but the displayed image is just a noise image without any meaningful information.
